File: app/src/main/java/com/cilenco/skiptrack/MediaSessionManagerProxy.java
package com.cilenco.skiptrack;

import android.annotation.SuppressLint;
import android.media.session.MediaSessionManager;
import android.os.Handler;
import android.view.KeyEvent;

import androidx.annotation.NonNull;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@SuppressWarnings("JavaReflectionMemberAccess")
@SuppressLint("PrivateApi")
public class MediaSessionManagerProxy {

    private final MediaSessionManager mediaSessionManager;

    private final Method setOnVolumeKeyLongPressListener;
    private final Method dispatchVolumeKeyEvent;

    public MediaSessionManagerProxy(MediaSessionManager mediaSessionManager) {
        this.mediaSessionManager = mediaSessionManager;
        try {
            final Class<?> onVolumeKeyLongPressListenerClass = Class.forName("android.media.session.MediaSessionManager$OnVolumeKeyLongPressListener");
            setOnVolumeKeyLongPressListener = MediaSessionManager.class.getMethod("setOnVolumeKeyLongPressListener", onVolumeKeyLongPressListenerClass, Handler.class);

            dispatchVolumeKeyEvent = MediaSessionManager.class.getMethod("dispatchVolumeKeyEvent", KeyEvent.class, Integer.class, Boolean.class);
        } catch (ClassNotFoundException | NoSuchMethodException e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    public MediaSessionManager get() {
        return mediaSessionManager;
    }

    public void setOnVolumeKeyLongPressListener(Object listener, Handler handler) {
        try {
            setOnVolumeKeyLongPressListener.invoke(mediaSessionManager, listener, handler);
        } catch (IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    public void dispatchVolumeKeyEvent(@NonNull KeyEvent keyEvent, int streamType, boolean musicOnly) {
        try {
            dispatchVolumeKeyEvent.invoke(mediaSessionManager, keyEvent, streamType, musicOnly);
        } catch (IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e.getMessage());
        }
    }
}

File: app/src/main/java/com/cilenco/skiptrack/MediaSessionManagerProxy.kt
package com.cilenco.skiptrack

import android.annotation.SuppressLint
import android.media.session.MediaSessionManager
import android.os.Handler
import android.view.KeyEvent
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.lang.reflect.Proxy

/**
 * [MediaSessionManager] proxy that exposes hidden methods using reflection, namely [setOnVolumeKeyLongPressListener].
 *
 * Use [MediaSessionManagerProxy.get] to retrieve the actual [MediaSessionManager] instance.
 */
@SuppressLint("PrivateApi")
class MediaSessionManagerProxy(
    private val mediaSessionManager: MediaSessionManager
) {

    private var setOnVolumeKeyLongPressListener: Method

    init {
        val onVolumeKeyLongPressListenerClass = Class.forName(
            "android.media.session.MediaSessionManager\$OnVolumeKeyLongPressListener"
        )

        setOnVolumeKeyLongPressListener = MediaSessionManager::class.java.getMethod(
            "setOnVolumeKeyLongPressListener", onVolumeKeyLongPressListenerClass, Handler::class.java
        )
    }

    fun createOnVolumeKeyLongPressListenerProxy(onVolumeKeyLongPress: (KeyEvent) -> (Unit)): Any {
        val clazz = Class.forName(
            "android.media.session.MediaSessionManager\$OnVolumeKeyLongPressListener"
        )

        return Proxy.newProxyInstance(clazz.classLoader, arrayOf(clazz)) { _: Any, _: Method, args: Array<Any> ->
            onVolumeKeyLongPress(args[0] as KeyEvent)
        }
    }

    fun get(): MediaSessionManager {
        return mediaSessionManager
    }

    /**
     * Set the volume key long-press listener. While the listener is set, the listener
     * gets the volume key long-presses instead of changing volume.
     *
     * <p>System can only have a single volume key long-press listener.
     *
     * @param listener The volume key long-press listener. {@code null} to reset.
     * @param handler The handler on which the listener should be invoked, or {@code null}
     *            if the listener should be invoked on the calling thread's looper.
     */
    fun setOnVolumeKeyLongPressListener(listener: Any?, handler: Handler?) {
        setOnVolumeKeyLongPressListener.invoke(mediaSessionManager, listener, handler)
    }
}